### PR TITLE
Don't charge for memory pages second time during a block execution

### LINKED ISSUE #1088: Charge for page load optionally
### Problem to Solve

At the moment, while processing queue (see `pallets/gear/fn process_queue`) we charge gas for loading programs pages every time, but this may be optimised due to overlay existence

### Possible Solution

To optimise with overlay:
Store program ids in ext_manager, which already were executed with current ext_manager instance: this means that pages were already read and/or updated, so the next read from db (overlay) is free.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by gshep**
Resolves #1088 

- don't execute functions without exports
- introduce additional gas value charged for the next time of loading memory pages during a block execution

@gear-tech/dev

### CODE REVIEW

File: core-processor/src/configs.rs [hunk 1]
```diff
@@ -53,6 +54,8 @@ pub struct AllocationsConfig {
     pub mem_grow_cost: u64,
     /// Load page cost.
     pub load_page_cost: u64,
+    /// Second load page cost.
```
Comment (grishasobol) [inline core-processor/src/configs.rs:57]: So, if the name is `second_` then we should explain in comment that it's not second only

File: core-processor/src/executor.rs [hunk 2]
```diff
@@ -93,9 +71,36 @@ fn make_checks_and_charge_gas_for_pages<'a>(
             return Err(ExecutionErrorReason::InitialMemoryGasExceeded);
         }
 
-        static_pages
+        return Ok(static_pages);
```
Comment (grishasobol) [inline core-processor/src/executor.rs:74]: Comment please

File: core-processor/src/executor.rs [hunk 3]
```diff
@@ -253,11 +260,33 @@ pub fn execute_wasm<A: ProcessorExt + EnvExt + IntoExtInfo + 'static, E: Environ
         }
     };
 
+    let is_initial = program.get_allocations().is_empty();
+
+    let exports = program.code().exports();
```
Comment (grishasobol) [inline core-processor/src/executor.rs:265]: Please check what can we do in core backend sandbox with exports then

**Review by grishasobol — APPROVED** [review]

**Comment by grishasobol** [comment]
FYI. As I can see in
https://github.com/gear-tech/substrate/blob/677fcc194cef7c670c03f7a2d596afedb6fd8c66/primitives/state-machine/src/ext.rs#L183-L207
if data is just read from storage, it's not copied to overlay. But it can be stored in DB hot caches, and also we won't have need to spend one page twice to relay chain. So, as conclusion, second+ loading should be cheaper

File: core-processor/src/executor.rs (hunk 2 above)
Comment (gshep) [reply core-processor/src/executor.rs:74]: Sorry didn't get what there should be commented.

Comment (grishasobol) [reply core-processor/src/executor.rs:74]: In case of init we return here and skip some below gas charging - explain in comment why

**Comment by breathx** [comment]
> FYI. As I can see in
> https://github.com/gear-tech/substrate/blob/677fcc194cef7c670c03f7a2d596afedb6fd8c66/primitives/state-machine/src/ext.rs#L183-L207
> if data is just read from storage, it's not copied to overlay. But it can be stored in DB hot caches, and also we won't have need to spend one page twice to relay chain. So, as conclusion, second+ loading should be cheaper

Seems that it happens in other places of code, cause our manual test with @gshep showed that multiple reads occurs as single read in benchmarks.

File: core-processor/src/executor.rs [hunk 4]
```diff
@@ -250,11 +254,33 @@ pub fn execute_wasm<A: ProcessorExt + EnvExt + IntoExtInfo + 'static, E: Environ
...
+    if !exports.contains(&dispatch.kind()) {
+        return Ok(DispatchResult {
```
Comment (breathx) [inline core-processor/src/executor.rs:261]: I would provide something like `DispatchResult::default(dispatch, pid, allocations)`

File: core-processor/src/executor.rs [hunk 5]
```diff
@@ -56,45 +57,46 @@ fn make_checks_and_charge_gas_for_pages<'a>(
     }
 
     let mem_size = if let Some(max_wasm_page) = allocations.iter().next_back() {
```
Comment (breathx) [inline core-processor/src/executor.rs:59]: in fact, charging in executor without lazy pages goes already after page loading

File: pallets/gear/src/manager/mod.rs [hunk 6]
```diff
@@ -146,14 +146,29 @@ where
...
     }
 
     /// NOTE: By calling this function we can't differ whether `None` returned, because
     /// program with `id` doesn't exist or it's terminated
-    pub fn get_actor(&self, id: ProgramId, with_pages: bool) -> Option<Actor> {
+    pub fn get_actor(
+        &mut self,
+        id: ProgramId,
+        with_pages: bool,
+        subsequent_execution: bool,
```
Comment (breathx) [inline pallets/gear/src/manager/mod.rs:170]: useless field here, I suppose

File: pallets/gear/src/tests.rs [hunk 7]
```diff
@@ -76,6 +76,20 @@ fn unstoppable_block_execution_works() {
...
+        let GasInfo {
+            burned: subsequent_burned_gas,
+            ..
+        } = Gear::calculate_gas_info(
+            USER_1.into_origin(),
+            HandleKind::Handle(program_id),
+            EMPTY_PAYLOAD.to_vec(),
+            0,
+            true,
+            true,
```
Comment (breathx) [inline pallets/gear/src/tests.rs:92]: I see no reason to add this field to rpc call. We have no guarantees when message will be executed, so to calculate it properly, need to add this value in min_limit by default and provide some filed like `might_be_returned` with amount of gas which might not be burned due to some conditions

File: pallets/gear/src/manager/mod.rs [hunk 8]
```diff
@@ -177,12 +192,16 @@ where
             Default::default()
         };
 
+        let subsequent_execution = subsequent_execution || self.programs.contains(&id);
+        self.programs.insert(id);
```
Comment (breathx) [inline pallets/gear/src/manager/mod.rs:196]: being in self.programs is not guarantee that we were executing this program this block: if we check some user id and its a program, we also append the id in programs

Comment (breathx) [reply pallets/gear/src/manager/mod.rs:196]: one more thing - if we try to execute program and have no gas to load pages also means that further execution should try to do this too

**Review by breathx — CHANGES_REQUESTED** [review]

File: core-processor/src/executor.rs (hunk 4 above)
Comment (gshep) [reply core-processor/src/executor.rs:261]: Agree. Moreover I move this change to the standalone PR siince there are some discussions around memory pages charging.

Comment (gshep) [reply core-processor/src/executor.rs:261]: done - #1183

**Pull request closed without merging**